apps/ejabberd/src/ejabberd_c2s.erl

```diff
@@ -1568,12 +1592,24 @@ maybe_send_element_safe(State, El) ->
         _ -> error
     end.
 
-send_element(#state{server = Server, sockmod = SockMod} = StateData, El)
-  when StateData#state.xml_socket ->
-    ejabberd_hooks:run(xmpp_send_element, Server, [Server, El]),
+%% @doc This is the termination point - from here stanza is sent to the user
+%% We sent the original stanza ('element') unless there is a different thing
+% keyed 'to_send'
+-spec send_element(mongoose_acc:t() | state(), state() | xmlel()) -> mongoose_acc:t() | ok.
+send_element(#state{} = StateData, #xmlel{} = El) ->
+    ?DEPRECATED,
+    send_element(mongoose_acc:from_element(El), StateData),
+    ok;
+send_element(Acc, #state{server = Server} = StateData) ->
+    Acc1 = ejabberd_hooks:run_fold(xmpp_send_element, Server, Acc, [Server]),
```

Comment:
Yes and no - there is a 'server' key in Acc, but it is taken from state of the sender's c2s; the Acc is then routed to the recipient's c2s which if I get it right may have a different domain. So, we can do it but first we'd have to store the other domain to and differentiate between them. Actually if we go this way we probably should store most of the data separately for sender and recipient (hooks run, privacy check etc); or go the other way and treat an acc entering the recipient's c2s as something brand new.